### ISSUE TITLE
When using ocLazyLoad RequireJs tryis to load files, taken from define, without .js ending

### ISSUE THREAD

**Issue by vytautas-pranskunas-** [ocombe/ocLazyLoad#144]
Hello, i have simple ocLazyLoad setup:

```
app = angular.module(appName, dependencies)
    .config(($ocLazyLoadProvider) => {
        $ocLazyLoadProvider.config({
            asyncLoader: require,
        });
    });

angular.bootstrap(document.getElementsByTagName("body")[0], [appName]);
```

then i have state:

```
$stateProvider.state('app.sand-box',
            {
                url: '/sand-box',
                controller: 'SandBoxController',
                views: {
                    'content@': {
                        templateProvider: ['$clientTemplate', ($clientTemplate: IClientTemplateService) => $clientTemplate.get('Web/SandBox/Views/sandbox.cshtml')]
                    }
                },
                resolve: {
                    store: $ocLazyLoad => $ocLazyLoad.load("/Web/Sandbox/SandBox.js")
                }
            });
```

and when ocLazy loads SandBox.js which look like:

```
define(["require", "exports", './Controllers/SandBoxController'], function(require, exports, impSC) {
 
    angular.module('SandBox', []).controller(impSC.SandBoxController.controllerId, impSC.SandBoxController);

});
```

network shows 404 because of /Controllers/SandBoxController but not /Controllers/SandBoxController.js was tried to be loaded. This is not happening in regular usage of angularjs.

Do you know why this issue may occur? 

Thanks

**Comment by ocombe** [comment]
Hmm it looks like the issue comes from RequireJS, not ocLazyLoad. Since you use it as your loader, ocLazyLoad will not load the files directly but let RequireJS do the work.

I'm not an expert of RequireJS, but if you change SandBox.js code to:
```
define(["require", "exports", './Controllers/SandBoxController.js'], function(require, exports, impSC) {

    angular.module('SandBox', []).controller(impSC.SandBoxController.controllerId, impSC.SandBoxController);

});
```

It should work ? (I just added .js to the url for your controller)

**Comment by vytautas-pranskunas-** [comment]
Yes if change it by adding .js it works but thing is that we use TypeScript and its AMD importing so it generates javascript files for us so import with .js is not valid. But as i mentioned there is no such issue without ocLazyLoad and i never saw this kind of issue with RequireJs because you have to install third party libraries to disable RequireJs abilty to add .js ending. So I think anyway this comes because of ocLazyLoad. Is there eny chance for you to try just add those few simple configs along with requirejs and any controllers maybe you will quickly find an issu.

**Comment by ocombe** [comment]
I'm working on a full rewrite of the lib to make it more modular. One of the big changes will be that you can call the injector on its own, so you should be able to use RequireJS and then call the injector (but you will have to give it the name of the module that you're loading).
Another change will be that the loaders will be separated from the core, but I'm not sure if it will fix your problem though.

Anyway like I said, I don't use RequireJS so.. @amcdnl might be more suited to help you there.

But if you could make me a plunkr with a minimal test case, that would help :)

**Comment by vytautas-pranskunas-** [comment]
I have just googled another topic like this please take a look:

https://groups.google.com/forum/#!msg/requirejs/9fD2bh5aoic/2j0yPKE7klMJ

so this should be really an issue. 
You said that you do not work with RequireJs but how does your library provde "asyncLoader: require" is there any body else who does? maybe he can help?

**Comment by vytautas-pranskunas-** [comment]
You have talked about same issue on https://github.com/ocombe/ocLazyLoad/issues/50
I think this might happen before i have added '/' before web 

```
 store: $ocLazyLoad => $ocLazyLoad.load("/Web/Sandbox/SandBox.js")
```

i will check tomorrow.

**Comment by ocombe** [comment]
Let me know if that's the problem :)

It works with jasmine but you'll have to trigger the digests because of the promises, you can check the test folder to see how I do it.
It works with ui router as well, but I'm not sure what ui router state tests are :)

**Comment by vytautas-pranskunas-** [comment]
Yes that was a problem with RequireJs.
But i have problem when testing ui.router states. 

Registered state:
```
 $stateProvider.state('app.sand-box',
            {
                url: '/sand-box',
                templateUrl:'Web/SandBox/Views/sandbox.cshtml'),
                resolve: {
                    store: $ocLazyLoad => $ocLazyLoad.load("../Web/Sandbox/SandBox.js")
                }
            });
```
Test:
```
 it("should route to sandbox page",function(){
            $state.go('app.sand-box');
            $scope.$digest();
            expect($state.current.url).toBe('/sand-box');
});
```

Cases:
1) When there is no resolve with $ocLazyLoad in state everything works fine.
2) When there is $ocLazyLoad it does not work at all - url does not chage
3) When i call $scope.$apply(); instead of $scope.$digest(); i got redirected to default state - so basically it means that it was not able to resolve Resolve section

Do you have any ideas, examples? Maybe i have to create some spy to avoid loading real file or something?

**Comment by ocombe** [comment]
The problem is that I use multiple promises in my lib and only one digest won't be enough to resolve all of them, that's probably why nothing changes in your test, it didn't resolved all the way up the router.

In my test I use a setInterval that will trigger digest every 50ms or so and I stop that setInterval once all promises have been resolved.
It's dirty and I don't like it, but I haven't found a better way to do it. If you find one, please let me know. Perhaps I should try to use less $q promises in my lib to fix that...

**Comment by vytautas-pranskunas-** [comment]
Yes I think you should do something to reduce promises because that's is an issue not only for tests but for performance too. Also one $digest should be enough for angular to resolve everything. However for now i have found dirty way round

```
spyOn($ocLazyLoad, 'load').and.returnValue('');
```

I have just mocked your call and do not wait for response. This how I ignore promises your library creates.

**Comment by ocombe** [comment]
Yes, I'll try to see if I can remove them. Registering components with angular should be synchronous anyway, it's just a matter of using callbacks instead of promises.

**Comment by vytautas-pranskunas-** [comment]
Please let me know when you make some changes so i can update. Thanks.

**Comment by mkoczorowski** [comment]
I don't think asyncLoader: require, is a valid config here

**Comment by folterung** [comment]
@vytautas-pranskunas- I ran into this same issue. I was able to use the noext plugin to allow me to load files with their extension and this fixed the conflict between requirejs and oclazyload.
